File: backend/app/Models/SupportTicket.php
<?php

namespace App\Models;

use App\Enums\SupportTicketRequesterRole;
use App\Enums\SupportTicketStatus;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\MorphMany;

class SupportTicket extends Model
{
    public $incrementing = false;

    protected $keyType = 'string';

    protected function casts(): array
    {
        return [
            'requester_role' => SupportTicketRequesterRole::class,
            'status' => SupportTicketStatus::class,
        ];
    }

    protected $fillable = [
        'id',
        'ticket_number',
        'requester_user_id',
        'requester_role',
        'category',
        'priority',
        'status',
        'subject',
        'description',
        'assigned_to_admin_id',
        'related_order_id',
        'related_product_id',
        'first_response_at',
        'resolved_at',
        'satisfaction_rating',
        'satisfaction_comment',
    ];

    /**
     * Requester is a Customer (requester_role = 'customer')
     * or a Vendor (requester_role = 'seller').
     * Use the relevant helper based on requester_role.
     */
    public function requesterCustomer(): BelongsTo
    {
        return $this->belongsTo(Customer::class, 'requester_user_id');
    }

    public function requesterVendor(): BelongsTo
    {
        return $this->belongsTo(Vendor::class, 'requester_user_id');
    }

    public function requesterTravelAgency(): BelongsTo
    {
        return $this->belongsTo(TravelAgency::class, 'requester_user_id');
    }

    public function assignedToAdmin(): BelongsTo
    {
        return $this->belongsTo(Admin::class, 'assigned_to_admin_id');
    }

    public function relatedOrder(): BelongsTo
    {
        return $this->belongsTo(Order::class, 'related_order_id');
    }

    public function relatedProduct(): BelongsTo
    {
        return $this->belongsTo(Product::class, 'related_product_id');
    }

    public function messages(): HasMany
    {
        return $this->hasMany(TicketMessage::class, 'ticket_id');
    }

    public function files(): MorphMany
    {
        return $this->morphMany(File::class, 'model');
    }
}

File: delivery_app/lib/shared/models/support_ticket.dart
class TicketMessage {
  final int id;
  final String? message;
  final DateTime? createdAt;

  TicketMessage({required this.id, this.message, this.createdAt});

  factory TicketMessage.fromJson(Map<String, dynamic> json) => TicketMessage(
        id: (json['id'] as num).toInt(),
        message: json['message'] as String?,
        createdAt: json['created_at'] != null ? DateTime.tryParse(json['created_at']) : null,
      );
}

class SupportTicket {
  final int id;
  final String ticketNumber;
  final String? category;
  final String? priority;
  final String? status;
  final String? subject;
  final int? relatedAssignmentId;
  final DateTime? createdAt;
  final DateTime? resolvedAt;
  final int? satisfactionRating;
  final String? satisfactionComment;
  final List<TicketMessage> messages;

  SupportTicket({
    required this.id,
    required this.ticketNumber,
    this.category,
    this.priority,
    this.status,
    this.subject,
    this.relatedAssignmentId,
    this.createdAt,
    this.resolvedAt,
    this.satisfactionRating,
    this.satisfactionComment,
    this.messages = const [],
  });

  factory SupportTicket.fromJson(Map<String, dynamic> json) => SupportTicket(
        id: (json['id'] as num).toInt(),
        ticketNumber: json['ticket_number'] as String? ?? '',
        category: json['category'] as String?,
        priority: json['priority'] as String?,
        status: json['status'] as String?,
        subject: json['subject'] as String?,
        relatedAssignmentId: (json['related_assignment_id'] as num?)?.toInt(),
        createdAt: json['created_at'] != null ? DateTime.tryParse(json['created_at']) : null,
        resolvedAt: json['resolved_at'] != null ? DateTime.tryParse(json['resolved_at']) : null,
        satisfactionRating: (json['satisfaction_rating'] as num?)?.toInt(),
        satisfactionComment: json['satisfaction_comment'] as String?,
        messages: (json['messages'] as List? ?? [])
            .map((e) => TicketMessage.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}
